### PR TITLE
NEW Create action

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripeltd/product-issues/issues/753

Example of it in action - https://github.com/emteknetnz/silverstripe-config/commits/2

Here's an example of it failing on merge-conflict - https://github.com/emteknetnz/silverstripe-config/actions/runs/5526914815/jobs/10082502001

Here's an example of it failing where there's a difference in composer.json - https://github.com/emteknetnz/silverstripe-config/actions/runs/5527199168/jobs/10082717263

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,14 @@
-# gha-merge-up
-GitHub Action to merge-up supported branches in a repository
+# GitHub Actions - Merge-up
+
+Merge-up supported branches in a repository
+
+## Usage
+
+**workflow.yml**
```
Comment (ChloeSartorelli) [inline README.md]:
```suggestion
**.github/workflows/merge-up.yml**
```
May as well be specific - this mirrors the example in the gha-ci readme.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,14 @@
...
+```yml
+steps:
+  - name: Merge-up
+    uses: silverstripe/gha-merge-up@v1
```
Comment (ChloeSartorelli) [inline README.md:11]: This doesn't seem correct? Doesn't it need to be in a structure like
```yml
jobs:
  merge-up:
    name: Merge-up
    steps:
        name: Merge-up
        uses: silverstripe/gha-merge-up@v1
```

And since this is useless if not on a cron, we may as well provide the full example:
```suggestion
on:
  # Run on two consecutive days of the week
  schedule:
  - cron: '0 0 * * 1,2'

jobs:
  merge-up:
    name: Merge-up
    # Only run the cron on the account hosting this repository, not on the accounts of forks
    # Change '<account_name>' to match the name of the account hosting this repository
    if: (github.event_name == 'schedule' && github.repository_owner == '<account_name>') || (github.event_name != 'schedule')
    runs-on: ubuntu-latest
    steps:
      - name: Merge-up
        uses: silverstripe/gha-merge-up@v1
```

File: action.yml [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+      run: |
+        # The minimum cms major with commercial support - configured at a global level
+        # Change this when major version support changes
+        MINIMUM_CMS_MAJOR=4
+
+        # Get the default branch from GitHub API
+        # We need to make an API call rather than just assume that the current branch is the default
+        # because this workflow may be triggered by workflow_dispatch on any branch
+        RESP_CODE=$(curl -w %{http_code} -s -o __base.json \
+        -X GET "https://api.github.com/repos/$GITHUB_REPOSITORY" \
```
Comment (ChloeSartorelli) [inline action.yml:29]: Please add a link to the relevant docs in a comment

File: action.yml [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        -H "Accept: application/vnd.github+json" \
+        -H "Authorization: Bearer ${{ github.token }}" \
+        -H "X-GitHub-Api-Version: 2022-11-28" \
+        )
+        if [[ $RESP_CODE != "200" ]]; then
+          echo "Unable to read list of tags - HTTP response code was $RESP_CODE"
+          exit 1
+        fi
+
+        # Gets all tags from GitHub API
```
Comment (ChloeSartorelli) [inline action.yml]:
```suggestion
        # Gets 100 most recently created tags from GitHub API
```

File: action.yml [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        fi
+
+        DEFAULT_BRANCH=$(jq -r .default_branch __base.json)
+        echo "DEFAULT_BRANCH is $DEFAULT_BRANCH"
+        rm __base.json
+
+        if [[ $DEFAULT_BRANCH != $GITHUB_REF_NAME ]]; then
+          echo "Current branch $GITHUB_REF_NAME is not the same as default branch $DEFAULT_BRANCH"
+          exit 1
+        fi
```
Comment (ChloeSartorelli) [inline action.yml]: Should probably move this to above the tag fetching call - no sense in fetching tags from the API if we're not doing anything on this branch anyway.

File: action.yml [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+        DEFAULT_BRANCH=$(jq -r .default_branch __base.json)
+        echo "DEFAULT_BRANCH is $DEFAULT_BRANCH"
+        rm __base.json
+
+        if [[ $DEFAULT_BRANCH != $GITHUB_REF_NAME ]]; then
+          echo "Current branch $GITHUB_REF_NAME is not the same as default branch $DEFAULT_BRANCH"
+          exit 1
```
Comment (ChloeSartorelli) [inline action.yml]: This won't work for various modules as defined on https://github.com/silverstripe/.github/issues/37#issuecomment-1524292415 and won't work for CMS 6 until we change the default branch - which probably won't be until after it goes stable, meaning we'd have to do manual merge-ups for it.

Might be more robust to get a list of branches (which we're doing above anyway), and check that we're on the latest that matches `^\d$` - since I don't think it matters so much what the _default_ branch is, as it matters what the _highest major branch_ is.

If doing that, move this logic and the branches fetch to be the first part of the step, so we can skip _everything_ else if we're not on the right branch.

File: action.yml [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        # Gets all branches from GitHub API
+        # https://docs.github.com/en/rest/branches/branches?apiVersion=2022-11-28#list-branches
+        RESP_CODE=$(curl -w %{http_code} -s -o __branches.json \
+        -X GET "https://api.github.com/repos/$GITHUB_REPOSITORY/branches?per_page=100" \
+        -H "Accept: application/vnd.github+json" \
+        -H "Authorization: Bearer ${{ github.token }}" \
+        -H "X-GitHub-Api-Version: 2022-11-28" \
+        )
+        if [[ $RESP_CODE != "200" ]]; then
+          echo "Unable to read list of tags - HTTP response code was $RESP_CODE"
```
Comment (ChloeSartorelli) [inline action.yml]:
```suggestion
          echo "Unable to read list of branches - HTTP response code was $RESP_CODE"
```

File: action.yml [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        # We need to make an API call rather than just assume that the current branch is the default
+        # because this workflow may be triggered by workflow_dispatch on any branch
+        RESP_CODE=$(curl -w %{http_code} -s -o __base.json \
+        -X GET "https://api.github.com/repos/$GITHUB_REPOSITORY" \
+        -H "Accept: application/vnd.github+json" \
+        -H "Authorization: Bearer ${{ github.token }}" \
+        -H "X-GitHub-Api-Version: 2022-11-28" \
+        )
+        if [[ $RESP_CODE != "200" ]]; then
+          echo "Unable to read list of tags - HTTP response code was $RESP_CODE"
```
Comment (ChloeSartorelli) [inline action.yml]:
```suggestion
          echo "Unable to fetch repository meta data - HTTP response code was $RESP_CODE"
```

File: funcs.php [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+    // work out default major
+    preg_match('#^([0-9]+)+\.?[0-9]*$#', $defaultBranch, $matches);
+    $defaultMajor = $matches[1];
+    
+    // read __composer.json of the current (default) branch
+    $contents = $composerJson ?: file_get_contents('__composer.json');
+    $json = json_decode($contents);
+    if (is_null($json)) {
+        throw new Exception('Could not parse __composer.json');
```
Comment (ChloeSartorelli) [inline funcs.php]: Use [`json_last_error()`](https://www.php.net/manual/en/function.json-last-error.php) here - will speed up debugging.

File: funcs.php [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    }
+    $defaultCmsMajor = '';
+    $version = preg_replace('#[^0-9\.]#', '', $json->require->{'silverstripe/framework'} ?? '');
+    if (preg_match('#^([0-9]+)+\.?[0-9]*$#', $version, $matches)) {
+        $defaultCmsMajor = $matches[1];
```
Comment (ChloeSartorelli) [inline funcs.php]: Maybe check a couple of others (`silverstripe/admin` and `silverstripe/cms`) before falling back to using the PHP constraint.

File: funcs.php [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        $phpVersion = $json->require->{'php'} ?? '';
+        if (substr($phpVersion,0, 4) === '^7.4') {
+            $defaultCmsMajor = 4;
+        } elseif (substr($phpVersion,0, 4) === '^8.1') {
+            $defaultCmsMajor = 5;
+        }
+    }
+    if ($defaultCmsMajor === '') {
+        throw new Exception('Could not work out what the default CMS major version this module uses');
+    }
```
Comment (ChloeSartorelli) [inline funcs.php]: Can we reuse the logic that does this from gha-generate-matrix? It's pretty robust and would be a pain to maintain effectively the same logic in two different forms in two different places.

File: funcs.php [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        // filter out non-standard branches
+        if (!preg_match('#^([0-9]+)+\.?[0-9]*$#', $branch, $matches)) {
+            continue;
+        }
+        // filter out majors that are too old
+        $major = $matches[1];
+        if (($major + $majorDiff) < $minimumCmsMajor) {
+            continue;
+        }
+        // filter out minor branches that are pre-stable
```
Comment (ChloeSartorelli) [inline funcs.php]:
```suggestion
        // filter out minor branches that are pre-stable or so old their tags don't appear in the API call
```

File: funcs.php [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+    $minorsWithStableTags = [];
+    $contents = $tagsJson ?: file_get_contents('__tags.json');
+    foreach (json_decode($contents) as $row) {
+        $tag = $row->name;
+        if (!preg_match('#^([0-9]+)\.([0-9]+)\.([0-9]+)$#', $tag, $matches)) {
+            continue;
+        }
+        $minor = $matches[1] . '.' . $matches[2];
+        $minorsWithStableTags[] = $minor;
```
Comment (ChloeSartorelli) [inline funcs.php]:
```suggestion
        $minorsWithStableTags[$minor] = $minor;
```
There are gonna be a bunch of tags for each minor - cleaner to only include each minor once.

File: funcs.php [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        if (preg_match('#^([0-9]+)\.([0-9]+)$#', $branch, $matches)) {
+            $minor = $matches[1] . '.' . $matches[2];
+            if (!in_array($branch, $minorsWithStableTags)) {
+                continue;
+            }
+        }
+        // suffix a .999 minor version to major branches
+        if (preg_match('#^[0-9]+$#', $branch)) {
+            $branch .= '.999';
+        }
```
Comment (ChloeSartorelli) [inline funcs.php]: Please add an explanation of _why_ we're doing this to the code comment. I'm _guessing_ it's for sorting? But let's be nice to our future selves and not keep them guessing.

File: funcs.php (hunk 12 above)
Comment (ChloeSartorelli) [inline funcs.php]: _why_ are we filtering out pre-stable minor branches? Surely we want to include our bug fixes in the next minor release? And we _definitely_ don't want to skip for example the `5.1` branch and merge `5.0` straight to `5`.

Obviously including pre-stable branches will require changes to the "remove everything except the latest minor from each major line" section.

File: funcs.php [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    // sort so that newest is first
+    usort($branches, 'version_compare');
+    $branches = array_reverse($branches);
+    
+    // remove the .999
+    array_walk($branches, function(&$branch) {
+        $branch = preg_replace('#\.999$#', '', $branch);
+    });
+    
+    // remove everything except the latest minor from each major line
```
Comment (ChloeSartorelli) [inline funcs.php]:
```suggestion
    // remove everything except the latest minor and the major branch from each major line (e.g. remove 4.12, keep 4.13 and 4)
```

Comment (ChloeSartorelli) [reply funcs.php]: It's important that this keeps not just "4.13" like the current comment implies, but also keeps "4"

File: tests/BranchesTest.php
Comment (ChloeSartorelli) [inline tests/BranchesTest.php:1]: Please also add a test where we have minimum support for CMS 4, but we've created the relevant CMS 6 compat branch.
e.g merge ups like: 4.13 > 4 > 5.0 > 5 > 6

We're likely to start having CMS 6 branches pretty early, and it would suck for those to end up outdated like the old master branches did.

File: action.yml [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+          exit 0
+        fi
+        echo "branches=$BRANCHES" >> $GITHUB_OUTPUT
+        rm __tags.json
+        rm __branches.json
+        rm __composer.json
+
+        # Check to see if there is anything to merge-up
+        # Technically we could skip these API calls and just do the checking using only git, though
+        # doing this way not checkout the entire git history of the repo if possible
```
Comment (ChloeSartorelli) [inline action.yml]: > though doing this way not checkout the entire git history of the repo if possible

Can you please reword this part? I'm not sure what this means - and it seems pretty important given the downside we're accepting as a result.

File: action.yml [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+              echo "Unable to mergeup between $FROM_BRANCH and $INTO_BRANCH - there are changes in $DEPENDENCY_FILE"
+              exit 1
+            fi
+          done
+
+          # Don't allow merge-ups when there are JS changes that would require a yarn build
+          if [[ $(echo "$FILES" | grep client/) != "" ]]; then
+            echo "Unable to mergeup between $FROM_BRANCH and $INTO_BRANCH - there are changes to JS files"
+            exit 1
+          fi
```
Comment (ChloeSartorelli) [inline action.yml]: I don't think we should do this. If there's conflicts, it will fail on a conflict. If there's no conflicts, then there's no problem. I believe that's the consensus we reached during backlog refinement as well, which is why this wasn't added as an acceptance criterion.

Even if we did do this, we should only do it on `client/dist`.

```suggestion
```

File: action.yml [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+          # Continue if there's nothing to commit
+          if [[ "$GIT_STATUS" =~ 'nothing to commit, working tree clean' ]]; then
+            echo "No changes found when merging-up $FROM_BRANCH into $INTO_BRANCH. Skipping."
+            continue
+          fi
+
+          # Commit and push the merge-up
+          # The commit message matches the one created by git when doing a merge-up
+          # That only $FROM_BRANCH is quoted and not $INTO_BRANCH is intentional
+          git commit -m "Merge branch '$FROM_BRANCH' into $INTO_BRANCH"
```
Comment (ChloeSartorelli) [inline action.yml]:
```suggestion
          # --no-edit in the context of a merge commit uses the default auto-generated commit message.
          git commit --no-edit
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (emteknetnz) [reply README.md]: Updated

File: README.md (hunk 2 above)
Comment (emteknetnz) [reply README.md:11]: It's correct how I had it

I don't think we should include all the other stuff around it, the README.md shows that it's a step in a workflow, we shouldn't be defining the rest of a workflow here, it's assumed you already know how to do that - e.g. we only document the step here -https://github.com/silverstripe/gha-pull-request

File: action.yml (hunk 3 above)
Comment (emteknetnz) [reply action.yml:29]: Updated

File: action.yml (hunk 8 above)
Comment (emteknetnz) [reply action.yml]: Updated

File: action.yml (hunk 4 above)
Comment (emteknetnz) [reply action.yml]: Updated

File: action.yml (hunk 7 above)
Comment (emteknetnz) [reply action.yml]: Updated

File: action.yml (hunk 16 above)
Comment (emteknetnz) [reply action.yml]: Updated

File: action.yml (hunk 17 above)
Comment (emteknetnz) [reply action.yml]: I'm thinking we should keep this here for now simply because we have a [separate follow up card](https://github.com/silverstripe/.github/issues/77) to address JS, so that seems like a better place to deal with this block

File: action.yml (hunk 18 above)
Comment (emteknetnz) [reply action.yml]: Updated

File: funcs.php (hunk 9 above)
Comment (emteknetnz) [reply funcs.php]: Updated

File: funcs.php (hunk 10 above)
Comment (emteknetnz) [reply funcs.php]: Updated

File: funcs.php (hunk 11 above)
Comment (emteknetnz) [reply funcs.php]: We could, though it seems like overkill since everything just requires framework or cms - i.e. we need to copy in this https://github.com/silverstripe/gha-generate-matrix/blob/1/consts.php#L87

CI need to be more robust because it could be used by the community for use on any module. However gha-merge-ups is only for use on supported modules that we look after ... and basically everything just requires framework or cms

File: funcs.php (hunk 13 above)
Comment (emteknetnz) [reply funcs.php]: Updated

File: funcs.php (hunk 12 above)
Comment (emteknetnz) [reply funcs.php]: Updated

File: funcs.php (hunk 14 above)
Comment (emteknetnz) [reply funcs.php]: Updated

File: funcs.php (hunk 15 above)
Comment (emteknetnz) [reply funcs.php]: Updated

File: action.yml (hunk 6 above)
Comment (emteknetnz) [reply action.yml]: Assumption here is wrong

This workflow is only restricted to _run_ on the default branch - however we hit the GitHub API to get a list of all branches and then later on it will do a full history `git clone` (via `actions/checkout`) and then manually `git checkout` the branches that need to be merged-up

Restricting to the default branch is also good in that it protected against potentially weird stuff happening later, such as manually running this job in the future on an older branch (e.g. on CMS 5 when the default has moved on to CMS 6) - while the older branch may be say stuck on `v1` on this action while the newer branch uses `v2`

Default branch is also used in funcs.php to work out the minimum cms major to merge up from e.g. `4` - once we change the default branch to `6` then the default supported major becomes `5`

File: action.yml (hunk 5 above)
Comment (emteknetnz) [reply action.yml]: Updated

**Comment by emteknetnz** [comment]
@GuySartorelli Have implemented peer review feedback

Have tested updated code - works - https://github.com/emteknetnz/silverstripe-config/commits/2

File: tests/BranchesTest.php
Comment (emteknetnz) [reply tests/BranchesTest.php:1]: Have added test

File: funcs.php (hunk 12 above)
Comment (emteknetnz) [reply funcs.php]: Have removed and updated unit tests

File: README.md (hunk 2 above)
Comment (ChloeSartorelli) [reply README.md:11]: Given that we both have precedent for our preferred way of documenting this, I'm okay with it staying as is.

File: action.yml (hunk 17 above)
Comment (ChloeSartorelli) [reply action.yml]: I really disagree here. We explicitly discussed this in backlog refinement and decided to let these just fail due to merge conflicts. Again, if there's no conflict then there's no problem.

If you do feel strongly about keeping this, please at least only check for changes in `client/dist/` explicitly,

File: funcs.php (hunk 11 above)
Comment (ChloeSartorelli) [reply funcs.php]: I'm not saying copy, I'm saying reuse as in have it in some central place where we can just update logic or mapping once and it'll be reflected in both places. My main problem with this implementation is its reliance on PHP versions, which means we have _yet another_ thing we have to remember to update (or wait for it to break) next time we do a major release.

If there's no clean way to put that logic or mapping in a central location, then I'll accept this implementation.

File: action.yml (hunk 6 above)
Comment (ChloeSartorelli) [reply action.yml]: > Default branch is also used in funcs.php to work out the minimum cms major to merge up from e.g. 4 - once we change the default branch to 6 then the default supported major becomes 5

This is the part I'm talking about, really. I'm fine with having a mechanism to restrict to only run this on the default branch - but we can't safely assume that branch is _the most up to date branch_, and therefore we can't use it to identify the minimum CMS major to merge up from.

File: action.yml (hunk 17 above)
Comment (emteknetnz) [reply action.yml]: OK have changed to `client/dist`

**Comment by GuySartorelli** [comment]
There are still two unresolved conversations that you haven't responded to - please respond.

File: action.yml (hunk 6 above)
Comment (emteknetnz) [reply action.yml]: I'm not sure I follow

My thinking here is that this is the current state of play - it'll merge up from 4 to 5:
- CMS 5 << default branch
- CMS 4 << minimum supported

Once we create CMS 6 branches it will be this, it'll merge up from 4 to 5 to 6
- CMS 6
- CMS 5 << default branch
- CMS 4 << minimum supported

When CMS 6 is goes stable,  but before we've switched over all the default branches it will be this, and it'll merge up from 4 to 5 to 6 (more than we technically need to since 4 will be out of support at this point)
- CMS 6
- CMS 5 << default branch  << minimum supported
- CMS 4

When we switch over the default branches to CMS 6 it will be this, and it'll merge up from 5 to 6
- CMS 6 << default branch
- CMS 5 << minimum supported

File: funcs.php (hunk 11 above)
Comment (emteknetnz) [reply funcs.php]: It's not necessary IMO and adds a lot of complexity to this since we'd new too create another github-action just for doing determining the version, and then updating this script to use it, and then also updating gha-generate-matrix to also use it

**Comment by emteknetnz** [comment]
Replied

File: funcs.php (hunk 11 above)
Comment (ChloeSartorelli) [reply funcs.php]: Fair enough. I think it reduces complexity by having only one set of variables and logic for the same functionality that we need to maintain - but I won't press the point. This can stay as is.

File: action.yml (hunk 6 above)
Comment (ChloeSartorelli) [reply action.yml]: Right. But we have some modules where the default branch is _wrong_ right now. So they'll calculating the _wrong_ minimum supported branch _already_ as soon as we add this workflow to them - potentially merging up stuff from branches that we don't monitor because they're not supported branches.

Comment (emteknetnz) [reply action.yml]: I don't think that's really an issue?

Take for instance `symbiote/advanced-workflow` where the default branch `5` is CMS 4 instead of what it should be `6` for CMS 5

Since action.yml defines `MINIMUM_CMS_MAJOR=4` then it won't merge-up anything from CMS 3.  It'll only start from CMS 4.

I've just added a unit test that maps to the currently state of `symbiote/silverstripe-advanced-workflow` called `Module where default branch has not been changed from CMS 4 and there is a new CMS 6 branch` to prove that thinkgs work as they should i.e. `expected` is `['5.9', '5', '6.0', '6', '7']`

Comment (ChloeSartorelli) [reply action.yml]: Ahh, yup good point. I'd forgotten about that hardcoded value.
Thanks for adding the test as an additional safety net.

**Review by GuySartorelli — APPROVED** [review]
Looks good. Let's see how it goes on framework.
